apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -21,13 +28,136 @@ import { getQuestionCourse } from '../../lib/question-variant.js';
+  });
+
+  // Question prompt
+  messages.push({
+    role: 'user',
+    content: `Question: \n${question_prompt}`,
+  });
+
+  // Examples
+  for (const example of example_submissions) {
+    messages.push({
+      role: 'user',
+      content: `Example answer: \n${example.submission_text} \nGrade to this example answer: \n${example.score_perc}`,
+    });
+  }
+
+  // Student answer
+  messages.push({
+    role: 'user',
+    content: `Answer: \n${student_answer} \nHow would you grade this? Please return the json object.`,
+  });
+  return messages;
+}
+
+async function generateSubmissionEmbeddings({
+  course,
+  question,
+  assessment_question,
+  urlPrefix,
+  openai,
+}: {
+  question: Question;
+  course: Course;
+  assessment_question: AssessmentQuestion;
+  urlPrefix: string;
+  openai: OpenAI;
+}): Promise<number> {
+  const question_course = await getQuestionCourse(question, course);
+
+  const result = await queryRows(
+    sql.select_all_instance_questions,
+    {
+      assessment_id: assessment_question.assessment_id,
+      assessment_question_id: assessment_question.id,
+    },
+    InstanceQuestionSchema,
+  );
+
+  let newEmbeddingsCount = 0;
+
+  for (const instance_question of result) {
+    const { variant, submission } = await queryRow(
+      sql.select_last_variant_and_submission,
+      { instance_question_id: instance_question.id },
+      SubmissionVariantSchema,
+    );
+    const submission_embedding = await queryOptionalRow(
+      sql.select_embedding_for_submission,
+      { submission_id: submission.id },
+      SubmissionGradingContextEmbeddingSchema,
+    );
+
+    // Do not recalculate embedding if it already exists for a submission
+    if (submission_embedding) {
+      continue;
+    }
+    const urls = buildQuestionUrls(urlPrefix, variant, question, instance_question);
+    const questionModule = questionServers.getModule(question.type);
+    const render_submission_results = await questionModule.render(
+      { question: false, submissions: true, answer: false },
+      variant,
+      question,
+      submission,
+      [submission],
+      question_course,
+      urls,
+    );
+    const $ = cheerio.load(render_submission_results.data.submissionHtmls[0], null, false);
+    $('script').remove();
+    const student_answer = $.html();
+    const embedding = await createEmbedding(openai, student_answer, `course_${course.id}`);
+    await queryOneRowAsync(sql.update_embedding_for_submission, {
+      embedding: vectorToString(embedding),
+      submission_id: submission.id,
+      submission_text: student_answer,
+    });
```

Comment:
I think for now adding a column with the assessment question id would be very much doable. We might be able to think of better ways to put the delete option for embeddings rather than having a separated button (there are a lot of buttons already). 

One idea I can think of: in a future PR we will likely add a pop-up windows (/ "AI grading options" button) so the instructors can make customizations to the grading jobs with score range / number of examples to use... And we can put a "regenerate embedding while grading" as an option there. How does that sound?